---
title: "19_plot_PrediXcan_results_ord3CIPN8"
author: "Heather Wheeler"
date: "`r Sys.time()`"
output: html_document
---

Load packages, functions, paths
```{r load, warning=F,message=F}
library(dplyr)
library(tidyr)
library(ggplot2)
library(calibrate)
library(knitr)
"%&%" = function(a,b) paste(a,b,sep="")
my.dir = "/Volumes/dolan-lab/hwheeler/ThePlatinumStudy/GWAS/"
res.dir <- my.dir %&% "genotypes/UMich_imputation_results/PrediXcan/PrediXcan_results/"
source(my.dir %&% "genotypes/qqunif_px.r")
source(my.dir %&% "manhattan.R")
```

## ord3CIPN8 = pred_exp + agediagnosis 
### Bonf. sig. tissues:
- DGN-WB
- TW_Brain-Nucleusaccumbens-basalganglia_ElasticNet.0.5
- TW_Breast-MammaryTissue_ElasticNet.0.5
- TW_Cells-Transformedfibroblasts_ElasticNet.0.5
- TW_SmallIntestine-TerminalIleum_ElasticNet.0.5
- TW_Spleen_ElasticNet.0.5
- TW_Stomach_ElasticNet.0.5
- TW_WholeBlood_ElasticNet.0.5

```{r,warning=FALSE,fig.width=6,fig.height=6,eval=TRUE,results='asis'}
pheno <- 'ord3CIPN8'
tislist <- scan(my.dir %&% "pred_exp_list","c")
for(tis in tislist){
  res = read.table(res.dir %&% "N88_n953_ord3CIPN8_agediagnosis_" %&% tis %&% ".ordreg.PrediXcan",header=T)
  newres = mutate(res,CHR=as.numeric(gsub("chr","",chr)),BP=start,SNP=gene) %>% arrange(CHR,BP)
  res <- newres[complete.cases(newres),] #rm NAs
  res <- res[res$P != 0,]#rm P=0 (unstable, see permutations below)
  if(min(res$P)<0.0001){
    manhattan(res,suggestiveline=FALSE,genomewideline = -log10(0.05/dim(res)[1]),main=pheno %&% " " %&% tis,annotatePval = 0.0001,cex.main=0.8)
    tophits <- dplyr::filter(res,P<=0.0001) %>% arrange(P) %>% dplyr::select(-CHR,-BP,-SNP,-lCI, -uCI)
    cat('\n')
    print(kable(tophits))
    cat('\n')
  }
  else{
    manhattan(res,suggestiveline=FALSE,genomewideline = -log10(0.05/dim(res)[1]),main=pheno %&% " " %&% tis,cex.main=0.8)
    cat('\n')
  }
  qq<-qqunif(res$P,plot=TRUE,title=pheno %&% " " %&% tis)
  cat('\n')
}
```

###PrediXcan 10,000 perms select tissues
```{r,warning=FALSE,fig.width=6,fig.height=6,eval=TRUE,results='asis'}
tislist=c('DGN-WB','TW_Brain-Nucleusaccumbens-basalganglia_ElasticNet.0.5','TW_WholeBlood_ElasticNet.0.5')
tis="DGN-WB"
for(tis in tislist){
  overall<-read.table(res.dir %&% 'perms/N88_n953_ord3CIPN8_agediagnosis_' %&% tis %&% '.ordreg.PrediXcan.empP_permset1',header=T) %>% arrange(ensid)
  count.mat <- matrix(NA,nrow=dim(overall)[1],ncol=100)
  for(i in 1:100){
    res <- read.table(res.dir %&% 'perms/N88_n953_ord3CIPN8_agediagnosis_' %&% tis %&% '.ordreg.PrediXcan.empP_permset' %&% i,header=T) %>%   arrange(ensid)
    count.mat[,i]<-res$counta
  }
  empP <- rowSums(count.mat)/10000
  newoverall<-cbind(overall[,1:14],empP) %>% mutate(CHR=as.numeric(gsub("chr","",chr)),BP=start) %>% arrange(CHR,BP)
  print(ggplot(newoverall,aes(x=P,y=empP))+geom_point(alpha=1/5)+ggtitle(tis %&% ' PrediXcan 10k perms'))
  res <- newoverall %>% mutate(empP=ifelse(empP==0,1e-05,empP),SNP=gene) #replace empP=0 with 1e-05
  res <- res %>% dplyr::filter(P>0) #remove P=0 (unstable)
  res <- res[complete.cases(res),] #rm NAs
  kable(head(arrange(res,empP)))
  manhattan(res,suggestiveline=FALSE,genomewideline = -log10(0.05/dim(res)[1]),main="P ord3CIPN8 " %&% tis, annotatePval = 0.0001)
  qq<-qqunif(res$P,plot=TRUE,title="P ord3CIPN8 " %&% tis)
  empPres <- mutate(res,P=empP)
  manhattan(empPres,suggestiveline=FALSE,genomewideline = -log10(0.05/dim(empPres)[1]),main="empP ord3CIPN8 " %&% tis, annotatePval = 0.0001, ylim=c(0,5.2))
  qq<-qqunif(res$empP,plot=TRUE,title="empP ord3CIPN8 " %&% tis)
}
```